services/src/main/java/org/apache/druid/cli/CliPeon.java

```diff
@@ -205,7 +223,67 @@ public void configure(Properties properties)
   protected List<? extends Module> getModules()
   {
     return ImmutableList.of(
-        new DruidProcessingModule(),
+        Modules.override(new DruidProcessingModule()).with(
+            new Module()
+            {
+              @Override
+              public void configure(Binder binder)
+              {
+
+              }
+
+              @Provides
+              @ManageLifecycle
+              public QueryProcessingPool getProcessingExecutorPool(
+                  Task task,
+                  DruidProcessingConfig config,
+                  ExecutorServiceMonitor executorServiceMonitor,
+                  Lifecycle lifecycle
+              )
+              {
+                if (!task.supportsQueries()) {
+                  return new ForwardingQueryProcessingPool(Execs.dummy());
+                }
+                return new MetricsEmittingQueryProcessingPool(
+                    PrioritizedExecutorService.create(
+                        lifecycle,
+                        config
+                    ),
+                    executorServiceMonitor
+                );
```

Comment:
Do not use a `ForwardingQueryProcessingPool` since it is not meant to be used anyway.

```suggestion
                if (task.supportsQueries()) {
                    return super.getProcessingPoolExecutor(args);
                } else {
                    return () -> null;
                }
```

A similar simplification can be done for other methods too.

Reply:
null would look better than the direct processing pool since using the `DirectQueryProcessingPool.INSTANCE` looks wrong. IMO it means to do everything in the calling thread which isn't the expected behaviour. 

> since it is not meant to be used anyway

I didn't understand this part. Why should we not be using the `ForwardingQueryProcessingPool`. The benefit of my approach would be that the calling code wouldn't need to assume that the processing pool can be `null` anywhere, and handle that case separately. Moreover, it also acts as a safeguard in case any non querying task tries to submit a task to the pool, instead of complacently executing the task in the same thread (as with the direct processing pool).

Reply:
> I didn't understand this part. Why should we not be using the ForwardingQueryProcessingPool. The benefit of my approach would be that the calling code wouldn't need to assume that the processing pool can be null anywhere, and handle that case separately. Moreover, it also acts as a safeguard in case any non querying task tries to submit a task to the pool, instead of complacently executing the task in the same thread (as with the direct processing pool).

I meant that if we know upfront that this task is not meant to use the query processing pool, then we should _never_ return an instance that can be used at all, even if it causes the task to fail (since it was doing something illegal anyway).

I agree with your point about `null`.
How about we add a `NoopQueryProcessingPool` that throws `Unsupported` exception when anything is submitted to it?

Reply:
Also, I wish the `QueryProcessingPool` didn't extend `ListeningExecutorService`. Are the executor service methods ever called on the query processing pool?

Reply:
https://github.com/apache/druid/blob/2198001930c01f3da7efea3c3bad57b9e94cf370/processing/src/main/java/org/apache/druid/query/groupby/epinephelinae/GroupByMergingQueryRunner.java#L214

This is one usage of the processing pool as an executor service. Its javadoc also mentions such usages

https://github.com/apache/druid/blob/2198001930c01f3da7efea3c3bad57b9e94cf370/processing/src/main/java/org/apache/druid/query/QueryProcessingPool.java#L31-L34

I think a cleaner design would have been to have a method `getExecutor()` in the `QueryProcessingPool` interface. But since this is an `@ExtensionPoint`, I suppose we should leave it as is for now.

Reply:
> then we should never return an instance that can be used at all, even if it causes the task to fail

The `ForwardingQueryProcessingPool(Execs.dummy())` would do exactly that unless I am mistaken. The task would be delegated to the [dummy executor](https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/java/util/common/concurrent/DummyExecutorService.java#L35-L35) which throws UOE on any attempt to submit the task.
I attempted to create a `NoopQueryProcessingPool` while raising the PR, but it was doing the same thing. Maybe I can rename and make it clearer to read, or subclass the forwarding pool explicitly.

Reply:
> The ForwardingQueryProcessingPool(Execs.dummy()) would do exactly that unless I am mistaken. The task would be delegated to the [dummy executor](https://github.com/apache/druid/blob/master/processing/src/main/java/org/apache/druid/java/util/common/concurrent/DummyExecutorService.java#L35-L35) which throws UOE on any attempt to submit the task.

While this is true, there are small differences in using a dedicated `NoopQueryProcessingPool`:
- The intent is clearer to someone reading the code. Using the `Noop` implementation implies that it is meant to do nothing. Using a `Forwarding` pool with a dummy executor could mean that it is supposed to have partial functionality.
- The error message (and perhaps the stack trace too) would be more user-friendly. When using `Noop` pool, the exception is thrown by the processing pool itself rather than the underlying dummy executor service.

That said, this is not a blocker for this PR as it is a style choice really.
There are already some quirks of the `QueryProcessingPool` interface that could use some cleanup. We could address this then.